src/FSharp.Data.GraphQL.Client/Subscriptions.fs

```diff
@@ -0,0 +1,247 @@
+    let tryPickFirstAndKeepRemaining (source : IObservable<'T>) =
+        let sourceSeq = Observable.toSeq source
+        let first = Seq.tryHead sourceSeq
+        match first with
+        | Some first -> Some (first, Seq.skip 1 sourceSeq)
+        | None -> None
+
+    let data, deferred =
+        source
+        |> tryPickFirstAndKeepRemaining
+        |> Option.map (fun (first, remaining) ->
+            let deferred = {
+                new IObservable<JsonValue> with
+                    member __.Subscribe(observer) =
+                        remaining
+                        |> Seq.iter (function
+                            | Incoming payload -> observer.OnNext(payload)
+                            | Finished -> observer.OnCompleted())
+                        { new IDisposable with member __.Dispose() = () } }
+            match first with
+            | Incoming payload -> Some payload, deferred
+            | Finished -> None, deferred)
+        |> Option.defaultValue (None, Observable.empty)
+
+    interface IGraphQLSubscriptionResponse with
+        member __.Data = data
+        member __.Deferred = deferred
+        member __.Unsubscribe() = unsubscriber () |> Async.RunSynchronously
+        member __.AsyncUnsubscribe() = unsubscriber ()
+
+    interface IDisposable with
+        member __.Dispose() = unsubscriber () |> Async.RunSynchronously
+
+/// An interface for implementing subscription handlers for GraphQLProvider.
+type IGraphQLSubscriptionHandler =
+    inherit IDisposable
+    /// Wnen implemented, starts a connection asynchronously to a GraphQL server which supports subscriptions accordingly to the implementation protocol.
+    abstract member AsyncConnect : string * (string * obj) [] * bool -> Async<unit>
+    /// When implemented, start a subscription asynchronously to the GraphQL using the implementation protocol.
+    abstract member AsyncSubscribe : GraphQLSubscriptionRequest -> Async<IGraphQLSubscriptionResponse>
+    /// Wnen implemented, starts a connection to a GraphQL server which supports subscriptions accordingly to the implementation protocol.
+    abstract member Connect : string * (string * obj) [] * bool -> unit
+    /// When implemented, start a subscription to the GraphQL using the implementation protocol.
+    abstract member Subscribe : GraphQLSubscriptionRequest -> IGraphQLSubscriptionResponse
+    /// When implemented, closes the subscription connection to the GraphQL server using the implementation protocol.
+    abstract member Close : unit -> unit
+    /// When implemented, asynchronously closes the subscription connection to the GraphQL server using the implementation protocol.
+    abstract member AsyncClose : unit -> Async<unit>
+
+/// A GraphQL subscription handler which uses GraphQL Over Web Socket protocol.
+/// See https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md for information.
+type GraphQLOverWebSocketSubscriptionHandler() =
```

Comment:
This is the default implementation of the handler, which uses the spec we default to.